// json-api/build/test/unit/db-adapters/Mongoose/lib.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const mongoose = require("mongoose");
const chai_1 = require("chai");
const lib_1 = require("../../../../src/db-adapters/Mongoose/lib");
const APIError_1 = require("../../../../src/types/APIError");
const Errors = require("../../../../src/util/errors");
describe("Mongoose Adapter private helper functions", () => {
    describe("errorHandler", () => {
        const validationSchema = new mongoose.Schema({
            "required": { required: true, type: String },
            "number": { type: Number },
            "customSetter": {
                type: Boolean,
                set(v) {
                    if (v === 4) {
                        throw new APIError_1.default({ typeUri: "made-up-for-test" });
                    }
                    else if (v === 5) {
                        throw new Error("Generic error");
                    }
                    else {
                        return v !== 0;
                    }
                }
            }
        });
        const TestModel = mongoose.model("ValidationTest", validationSchema);
        const testModelErrors = (doc, testFn) => {
            return new TestModel(doc).validate().then(() => {
                throw new Error("Should not run");
            }, (e) => {
                try {
                    lib_1.errorHandler(e);
                }
                catch (result) {
                    return testFn(result, e);
                }
            });
        };
        it("should return a special error for missing required fields", () => {
            return testModelErrors({}, (errors, rawError) => {
                const error = errors[0];
                const rawRequiredError = rawError.errors.required;
                chai_1.expect(error).to.be.an.instanceof(APIError_1.default);
                chai_1.expect(error).to.deep.equal(Errors.missingField({
                    detail: rawRequiredError.message,
                    rawError: rawRequiredError
                }));
            });
        });
        it("should return a field invalid error w/ mongoose's message for built-in errors", () => {
            return testModelErrors({
                required: "present",
                number: "String!"
            }, (errors, rawError) => {
                const error = errors[0];
                const rawCastError = rawError.errors.number;
                chai_1.expect(error).to.be.an.instanceof(APIError_1.default);
                chai_1.expect(error).to.deep.equal(Errors.invalidFieldValue({
                    detail: 'Cast to Number failed for value "String!" at path "number"',
                    rawError: rawCastError
                }));
            });
        });
        it("should use generic message, but error.reason as rawError, if user threw custom error", () => {
            return testModelErrors({
                required: "present",
                customSetter: 5
            }, (errors, rawError) => {
                const error = errors[0];
                const rawCastError = rawError.errors.customSetter;
                chai_1.expect(error).to.be.an.instanceof(APIError_1.default);
                chai_1.expect(error).to.deep.equal(Errors.invalidFieldValue({
                    detail: 'Invalid value for path "customSetter"',
                    rawError: rawCastError.reason
                }));
            });
        });
        it("should use user-provided APIError if present", () => {
            return testModelErrors({
                required: "present",
                customSetter: 4
            }, (errors, rawError) => {
                const error = errors[0];
                chai_1.expect(error).to.be.an.instanceof(APIError_1.default);
                chai_1.expect(error).to.deep.equal(new APIError_1.default({ typeUri: "made-up-for-test" }));
            });
        });
        it("should support multiple errors at once", () => {
            return testModelErrors({
                customSetter: 4
            }, (errors, rawError) => {
                chai_1.expect(errors.some((it) => it.typeUri === "https://jsonapi.js.org/errors/missing-required-field")).to.be.true;
                chai_1.expect(errors.some((it) => it.typeUri === "made-up-for-test")).to.be.true;
            });
        });
    });
    describe("toMongoCriteria", () => {
        describe("predicates", () => {
            it("should return empty objects for predicates with empty arguments", () => {
                const or = { field: undefined, value: [], operator: "or" };
                const and = { field: undefined, value: [], operator: "and" };
                chai_1.expect(lib_1.toMongoCriteria(or)).to.deep.equal({});
                chai_1.expect(lib_1.toMongoCriteria(and)).to.deep.equal({});
            });
        });
        it("should support multiple operators for the same field", () => {
            const testPredicate = {
                operator: "and",
                value: [
                    { field: "field", value: "value", operator: "eq" },
                    { field: "field", value: "value", operator: "neq" },
                    { field: "field", value: "value", operator: "gte" },
                    {
                        operator: "or",
                        value: [
                            { field: "field", value: "value2", operator: "eq" },
                            { field: "field3", value: 1, operator: "gte" }
                        ],
                        field: undefined
                    }
                ],
                field: undefined
            };
            chai_1.expect(lib_1.toMongoCriteria(testPredicate)).to.deep.equal({
                $and: [
                    { field: "value" },
                    { field: { $ne: "value" } },
                    { field: { $gte: "value" } },
                    { $or: [{ field: "value2" }, { field3: { $gte: 1 } }] }
                ]
            });
        });
        it("should support multiple filters with the same operator + field", () => {
            chai_1.expect(lib_1.toMongoCriteria({
                operator: "and",
                value: [
                    { field: "id2", operator: "eq", value: "23" },
                    { field: "id2", operator: "in", value: ["1", "2"] },
                    { field: "id2", operator: "eq", value: "33" }
                ],
                field: undefined
            })).to.deep.equal({
                $and: [
                    { id2: "23" },
                    { id2: { $in: ["1", "2"] } },
                    { id2: "33" }
                ]
            });
        });
        it("should transform filters on id to filters on _id", () => {
            chai_1.expect(lib_1.toMongoCriteria({
                operator: "and",
                value: [{ field: "id", operator: "eq", value: "33" }],
                field: undefined
            })).to.deep.equal({
                $and: [
                    { _id: "33" }
                ]
            });
        });
        it("should support nested predicates", () => {
            const testPredicate = {
                operator: "and",
                value: [
                    { field: "field", value: "value", operator: "eq" },
                    {
                        operator: "or",
                        value: [
                            {
                                operator: "and",
                                value: [{ field: "field2", value: "varlu2", operator: "eq" }],
                                field: undefined
                            },
                            { field: "field3", value: 1, operator: "gte" }
                        ],
                        field: undefined
                    }
                ],
                field: undefined
            };
            chai_1.expect(lib_1.toMongoCriteria(testPredicate)).to.deep.equal({
                $and: [
                    { field: "value" },
                    {
                        $or: [
                            { $and: [{ field2: "varlu2" }] },
                            { field3: { $gte: 1 } }
                        ]
                    }
                ]
            });
        });
    });
});
